memc: stm32_xspi_psram: Fix compilation issues for STM32H5 series

tpambor opened the pull request:
The STM32H5 series has only a single XSPI interface and does not support the XSPI manager, the related `MaxTran` parameter and 16-bit IO. Configure these only if they are supported by the chip.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/99191

tpambor committed:
drivers: memc: stm32xspi_psram: only use xspi manager if supported by chip

The STM32H5 series has only a single XSPI interface and does not support
the XSPI manager or the related `MaxTran` parameter. Configure these
only if they are supported by the chip.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
drivers: memc: stm32xspi_psram: Use 16-bit mode only if supported by chip

XSPI on STM32H5 series does not support 16-bit mode. Guard relevant code
behind preprocessor checks to avoid compilation issues.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

erwango reviewed (changes requested):
Thanks. Btw, it is rather more adding driver support to H5 than fixing a mere bug. Also, it would be nice to be sure there is a configuration in tree allowing to build it. Can it be enabled on a board ? If not, would you mind providing a configuration in `tests/drivers/build_all`

JarmouniA reviewed (commented):
H5 has an OctoSPI controller (that's XSPI (JESD251C) standard compliant, which is not the same as having an XSPI controller like the N6 for ex.), that's why many things aren't available in the HAL.
To me, this driver should be checking the compatible of the parent of its device node, and decide what to use based on that.

etienne-lms reviewed (commented):
For consistency in commit message header lines, prefer
`drivers: memc: stm32xspi_psram: ...`

tpambor committed:
tests: drivers: build_all: memc: add test for stm32-xspi-psram

Add a new build test for the stm32-xspi-psram peripheral for the
stm32h573i_dk and stm32h7s78_dk boards.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor commented:
> Long term solution on this is to make use of upcoming MSPI API, which splits device and controllers implementations. See #96670 For short term, I think we need to find a trade of by getting the maximum out of the current driver but avoid to spend too much time on it.

Thanks for pointing this out. I wasn’t aware of that PR. The MSPI API approach definitely seems like the right direction. I will look into utilizing it to add support for additional PSRAMs.

erwango commented:
https://github.com/zephyrproject-rtos/zephyr/pull/96670 has now been merged.
